numba/unicode.py

```diff
@@ -440,6 +453,80 @@ def endswith_impl(a, b):
         return endswith_impl
 
 
+@overload_method(types.UnicodeType, 'split')
+def unicode_split(a, sep, maxsplit=-1):
+    if isinstance(sep, types.UnicodeType):
+        def split_impl(a, sep, maxsplit=-1):
+            a_len = len(a)
+            sep_len = len(sep)
+
+            if sep_len == 0:
+                raise ValueError('empty separator')
+
+            parts = []
+            last = 0
+            idx = 0
+            split_count = 0
+
+            while idx < a_len and (maxsplit == -1 or split_count < maxsplit):
+                if _cmp_region(a, idx, sep, 0, sep_len) == 0:
+                    parts.append(a[last:idx])
+                    idx += sep_len
+                    last = idx
+                    split_count += 1
+                else:
+                    idx += 1
+
+            if last <= a_len:
+                parts.append(a[last:])
+
+            return parts
+        return split_impl
+
+
+@njit
+def join_list(sep, parts):
+    parts_len = len(parts)
+    if parts_len == 0:
+        return ''
+
+    # Precompute size and char_width of result
+    sep_len = len(sep)
+    length = (parts_len - 1) * sep_len
+    kind = sep._kind
+    for p in parts:
+        length += len(p)
+        kind = _pick_kind(kind, p._kind)
+
+    result = _empty_string(kind, length)
+
+    # populate string
+    part = parts[0]
+    _strncpy(result, 0, part, 0, len(part))
+    dst_offset = len(part)
+    for idx in range(1, parts_len):
+        _strncpy(result, dst_offset, sep, 0, sep_len)
+        dst_offset += sep_len
+        part = parts[idx]
+        _strncpy(result, dst_offset, part, 0, len(part))
+        dst_offset += len(part)
+
+    return result
+
+
+@overload_method(types.UnicodeType, 'join')
+def unicode_join(sep, parts):
```

Comment:
Could the type of the content of `parts` be validated to ensure it is unicode else obscure TypingErrors appear from deep in the implementation detail?
e.g.
```python
from numba import njit

def foo():
    return 'z'.join([1,])

print('"%s"' % njit(foo)())
```
(error comes from trying to type `len()` applied to the integer)

and 

```python
from numba import njit

def foo():
    return 'z'.join([[1], [2]])

print('"%s"' % njit(foo)())
```
(error comes from `list(int64)` having not `_kind` attr).

Reply:
added this check.  The new typing error is still a little hard to understand, but better than blowing up on `len()`.  :)